### PR TITLE
common,log: use ISO 8601 datetime format

### PR DESCRIPTION
**Pull request by liewegas**
Make a hard change to ISO 8601 format.  That means,

- 'T' between date and time
- 'Z' suffix for GMT, or '+nnnn' or '-nnnn' for localtime to indicate timezone.
- fixes to parse to make everything work
- tests!

Anybody parsing our old ISO 8601-like output will probably need to be fixed.

Output is like ``2019-04-24T16:06:53.039991-0100`` for stringify, and ``2019-04-24T17:06:53.039991Z`` from the gmtime() method.

### CODE REVIEW

File: src/include/utime.h [hunk 1]
```diff
@@ -239,14 +239,14 @@ class utime_t {
       out << (long)sec() << "." << std::setw(6) << usec();
     } else {
       // this looks like an absolute time.
-      //  aim for http://en.wikipedia.org/wiki/ISO_8601
+      //  conform to http://en.wikipedia.org/wiki/ISO_8601
```
Comment (adamemerson) [inline src/include/utime.h]: You might want to use the routines in common/iso_8601.h

**Comment by LenzGr** [comment]
@a2batic does this PR supersede yours (#27770) in some way?

File: src/include/utime.h (hunk 1 above)
Comment (liewegas) [reply src/include/utime.h]: that'll work for gmtime(), but not the others.  the parser is tempting except it doesn't support the timezone portion.

**Comment by cbodley** [comment]
👍 i very much prefer this to dump_timestamp() in https://github.com/ceph/ceph/pull/27790

**Comment by a2batic** [comment]
> @a2batic does this PR supersede yours (#27770) in some way?

@LenzGr, yes, that PR is no longer required. I will go ahead and close it.

File: src/include/utime.h [hunk 2]
```diff
@@ -400,17 +381,45 @@ class utime_t {
 
     const char *p = strptime(date.c_str(), "%Y-%m-%d", &tm);
     if (p) {
-      if (*p == ' ') {
+      if (*p == ' ' || *p == 'T') {
```
Comment (cbodley) [inline src/include/utime.h]: rgw multisite is using utime_t::gmtime_nsec() to send timestamps between zones in json format. it looks like older zones will have trouble parsing the timestamps from octopus zones as a result. can we plan on backporting these utime_t::parse_date() changes to prepare for this mixed cluster case?

Comment (liewegas) [reply src/include/utime.h]: Yeah.  We'll need to document the minimum mimic and nautilus versions to run before doing the upgrade to octopus.

**Comment by cbodley** [comment]
@liewegas i'm still debugging the multisite failures. i haven't been able to reproduce them locally on fedora, but the teuthology runs on centos are consistently failing to decode utime_t in json:
```
2019-05-03T20:36:55.676+0000 34ff6700  0 ERROR: JSONDecoder failed with mtime: failed to decode utime_t
2019-05-03T20:36:55.678+0000 34ff6700  0 meta sync: ERROR: can't store key: user:realm-admin ret=-22
```

File: src/include/utime.h [hunk 3]
```diff
@@ -400,17 +381,45 @@ class utime_t {
...
+	    ++q;
+	  }
+	}
+	// look for tz...
+	if (*q == '-' || *q == '+' || *q == 'Z') {
+	  *q = '%';
+	  *(q+1) = 'z';
+	  *(q+2) = 0;
+	}
+	p = strptime(p, fmt, &tm);
```
Comment (cbodley) [inline src/include/utime.h]: i'm seeing this call fail on centos, using a simple unit test from commit https://github.com/cbodley/ceph/commit/0b6eecfbbb23130e3560b2b113d028966c5a5f11:
```
414             p = strptime(p, fmt, &tm);
(gdb) p p
$17 = 0x5555558acaf3 "16:53:20.181200Z"
(gdb) p fmt
$18 = "%H:%M:%S.181200%z", '\000' <repeats 14 times>
(gdb) n
415             if (!p) {
(gdb) p p
$19 = 0x0
```

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
jenkins retest this please

**Comment by liewegas** [comment]
@cbodley https://docs.google.com/document/d/1TJPNOBowwaGPVKMMuqvSmm_tgUeqnQr28MiNMQdwu-E/edit is the latest rgw run...

**Comment by cbodley** [comment]
@liewegas that looks like a pass to me! :+1:

the three-zone multisite test failures on test_versioned_object_incremental_sync are tracked in http://tracker.ceph.com/issues/39142

File: src/include/utime.h (hunk 2 above)
Comment (cbodley) [reply src/include/utime.h]: i'll work on backports under http://tracker.ceph.com/issues/40086

**Pull request merged**